Expand CONTRIBUTING.md (#3439)
DataDog/dd-trace-js: CONTRIBUTING.md

base. If you would like support for a module that is not listed, [contact support][1] to share a request.

[1]: https://docs.datadoghq.com/help

## Keep changes small and incremental

Changes should be incremental and understandable. As much as possible, large-scale efforts should be broken up into many PRs over time for better reviewability. If a feature would require more changes to be "complete" it's fine to land partial changes if they are not wired up to anything yet, so long as tests are included which at least prove those parts work in isolation.

There are great benefits to taking a measured and iterative approach to improvement. When working on code in fewer places there is far less risk of running into merge conflicts or incompatibilities with other systems. Keeping contributions small makes them easy to review which makes that much quicker to land. Additionally, keeping things small and iterative makes it easier for other teams to review and understand what the code does.

## Be descriptive!

> Sometimes code can be self-documenting, but often it can't. That is especially true to someone reviewing code they haven't worked on. If you're conscious of writing code in a self-describing way and leave comments anywhere that self-description fails you can go a long way to making even complex code coherent to one not already familiar with that code.

Reviewer: I feel this is a bit of a run-on sentence. I'd suggest:
```suggestion
Sometimes code can be self-documenting, but often it can't. That is especially true to someone reviewing code they haven't worked on. Be conscious of writing code in a self-describing way and leave comments anywhere that self-description fails. This goes a long way towards making even complex code coherent to one not already familiar with it.
```
This also turns it into an actionable instruction ("Do this. Here's why.") instead of a hypothetical ("if you did this".)

> Try to write code in a way the describes the intent when read. For example, verbs can be used for function and method names to communicate that they are used to do some specific action. In doing so it becomes clear when referenced by name elsewhere that it is a function and what the function is meant to do. If a function can not be described a simple verb it's probably too complex or does too many things.

Reviewer: ```suggestion
Try to write code in a way the describes the intent when read. For example, verbs can be used for function and method names to communicate that they are used to do some specific action. In doing so it becomes clear when referenced by name elsewhere that it is a function and what the function is meant to do. If a function can not be described with a simple verb it's probably too complex or does too many things.
```

## Avoid large refactors

Large refactors should generally be avoided in favour of iterative approaches. For example, rather than rewriting how every plugin works, one might make a special-case plugin that works a bit different for their particular use-case. If several dozen files need to change to add a feature we've probably done something wrong.

Sometimes new patterns or new ideas emerge which would be a substantial improvement over the existing state. It can be tempting to want to go all-in on a new way to do something, but the code churn can be hard to manage. It's best to introduce such new things incrementally and advocate for their adoption gradually through the rest of the codebase. As old systems are gradually phased out, the infrastructure which supports them can be deleted or relegated to lazy-loading only if and when that specific part of the system needs to be used.

## Always consider backportability

> To reduce delta between release lines and make it easier for us to support older versions we try as much as possible to backport every change we can. If we are diligent about keeping breaking changes to a minimum and ensuring we don't use language or runtime features which are too new we can generally be confident that a change can be backported.

Reviewer 2: Maybe we could add the fact that we prefere implementing breaking changes with a 
```javascript
if (version.major >= 4) { ... }
```
so we don't have code differences between lines

Author: If we minimize the scope of breaking changes we shouldn't actually _need_ to be putting version conditions like this everywhere. I would recommend putting the content of that block in a separate function and still backporting that, then the _major_ change is only the addition of calling that function.

Reviewer 2: No I know that's what I'm saying too. I was just saying it could be nice to have an example on how to code features for minimum release line conflicts. i.e.: implementing a feature but changing the default behavior, using major version checks etc...

Reviewer 2: Or maybe it's better to add those to the semver-major paragraph, up to you, just making suggestions here

Author: I made some improvements to that. Let me know what you think. 😄
